core/src/main/java/org/apache/iceberg/Partition.java

```diff
@@ -0,0 +1,355 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.Objects;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.apache.iceberg.types.Types;
+
+public class Partition implements IndexedRecord {
+  private PartitionData partitionData;
+  private int specId;
+  private long dataRecordCount;
+  private int dataFileCount;
+  private long dataFileSizeInBytes;
+  private long posDeleteRecordCount;
+  private int posDeleteFileCount;
+  private long eqDeleteRecordCount;
+  private int eqDeleteFileCount;
+  // Optional accurate count of records in a partition after applying the delete files if any
+  private long totalRecordCount;
+  // Commit time of snapshot that last updated this partition
+  private long lastUpdatedAt;
+  // ID of snapshot that last updated this partition
+  private long lastUpdatedSnapshotId;
+
+  public enum Column {
+    PARTITION_DATA,
+    SPEC_ID,
+    DATA_RECORD_COUNT,
+    DATA_FILE_COUNT,
+    DATA_FILE_SIZE_IN_BYTES,
+    POSITION_DELETE_RECORD_COUNT,
+    POSITION_DELETE_FILE_COUNT,
+    EQUALITY_DELETE_RECORD_COUNT,
+    EQUALITY_DELETE_FILE_COUNT,
+    TOTAL_RECORD_COUNT,
+    LAST_UPDATED_AT,
+    LAST_UPDATED_SNAPSHOT_ID
+  }
+
+  public Partition() {}
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public PartitionData partitionData() {
+    return partitionData;
+  }
+
+  public int specId() {
+    return specId;
+  }
+
+  public long dataRecordCount() {
+    return dataRecordCount;
+  }
+
+  public int dataFileCount() {
+    return dataFileCount;
+  }
+
+  public long dataFileSizeInBytes() {
+    return dataFileSizeInBytes;
+  }
+
+  public Long posDeleteRecordCount() {
+    return posDeleteRecordCount;
+  }
+
+  public Integer posDeleteFileCount() {
+    return posDeleteFileCount;
+  }
+
+  public Long eqDeleteRecordCount() {
+    return eqDeleteRecordCount;
+  }
+
+  public Integer eqDeleteFileCount() {
+    return eqDeleteFileCount;
+  }
+
+  public Long totalRecordCount() {
+    return totalRecordCount;
+  }
+
+  public Long lastUpdatedAt() {
+    return lastUpdatedAt;
+  }
+
+  public Long lastUpdatedSnapshotId() {
+    return lastUpdatedSnapshotId;
+  }
+
+  @Override
+  public void put(int i, Object v) {
```

Comment:
Extends `IndexedRecord` to use the existing Parquet/ORC Avro reader and writer from the `iceberg-parquet` and `iceberg-orc` module.